languages/python/django-oso/django_oso/auth.py

```diff
@@ -33,6 +38,65 @@ def authorize(request, resource, *, actor=None, action=None):
         raise PermissionDenied()
 
 
+def authorize_model(request, model, *, actor=None, action=None) -> Q:
+    """Authorize ``request`` for django model ``model``, ``actor``, and ``action``.
+
+    Partially evalutes the Polar rule allow(actor, action, Partial(model)). If
+    authorization fails, raises a :py:class:`django.core.exceptions.PermissionDenied`
+    exception.
+
+    Otherwise, returns a django ``Q`` object representing a filter that must be
+    applied to ``model``. This object can be applied to filter query results to
+    only contain authorized objects.
+
+    For example::
+
+        post_filter = authorize_model(request, Post)
+        authorized_posts = Post.objects.filter(post_filter)
+
+    :param actor: The actor making the request. Defaults to ``request.user``.
+    :param action: The action to authorize the actor to perform. Defaults to
+                    ``request.method``.
+    :param model: The model to authorize access for, or the string name of the model
+                  class in Polar (``app_name::ModelName``).
+
+    :raises django.core.exceptions.PermissionDenied: If the request is not authorized.
+    :returns: A django ``Q`` object representing the authorization filter.
+    """
+    if actor is None:
+        actor = request.user
+
+    if action is None:
+        action = request.method
+
+    try:
+        if issubclass(model, models.Model):
+            resource_type = get_model_name(model)
+    except TypeError:
+        assert isinstance(model, str)
+        resource_type = model
+
+    partial_resource = Partial("resource", TypeConstraint(resource_type))
+    results = Oso.query_rule("allow", actor, action, partial_resource)
+
+    filter = None
+    for result in results:
+        resource_partial = result["bindings"]["resource"]
+        if filter is None:
+            filter = Q()
+
+        next_filter = partial_to_query_filter(resource_partial, resource_type)
+        if next_filter == Q():
+            return next_filter
+
+        filter |= next_filter
```

Comment:
This is super cool if I understand it correctly.
If you had two different rules about why you could access something, those both become filters that are then or'd together in django?
Like if there's one rule that says you can see a widget if it's color is blue and another rule that says you can see a widget if it's color is green, you end up with a sql query that's like `select * from widgets where color = green or color = blue`?
Is that right?

Reply:
Yup, that is exactly right.

Reply:
🎆